Avoid resorting units when converting to strings

mhvk opened the pull request:
When converting units to strings, the order of the base units is now always the same as it is in the unit itself: by decreasing power to which the unit is raised, and then alphabetical. This avoids super-ugly representations such as `beam-1 Jy` (such as could happen for FITS; see #13217).

However, as can be seen from the PR, it also means some (doc)tests had to be adapted, and it can reasonably be foreseen that this will cause similar down-stream failures. ~Hence, opening as draft for now.~ EDIT: given initial positive feedback, changed to regular PR.

~(Note based on #14437 for now; will rebase once that is in.)~ EDIT: rebased now previous format PRs are in.

Fixes #13217

mhvk commented on astropy/units/format/generic.py:
```diff
@@ -594,11 +594,6 @@ def _do_parse(cls, s, debug=False):
                 else:
                     raise ValueError(f"Syntax error parsing unit '{s}'")
 
-    @classmethod
```
This is really the main change. Since FITS and VOUNIT inherit from `generic`, this means that they had the same sorting -- undoing the more logical sorting in their classes. Indeed, the sorting that was done in FITS/VOUNIT just followed what `CompositeUnit` itself does anyway, so in that sense for those formats this change makes them behave as intended...

pllim commented on astropy/units/tests/test_format.py:
```diff
@@ -659,7 +659,7 @@ def test_vounit_details():
         flam = u.erg / u.cm / u.cm / u.s / u.AA
         x = u.format.VOUnit.to_string(flam)
-        assert x == "Angstrom**-1.cm**-2.erg.s**-1"
+        assert x == "erg.Angstrom**-1.s**-1.cm**-2"
```
Yes, I like the new order. 😸

nstarman commented on docs/units/equivalencies.rst:
```diff
@@ -626,27 +626,27 @@ However, when passing the spectral equivalency, you can see there are
-    eV           | 1.60218e-19 kg m2 / s2 | electronvolt          ,
-    earthRad     | 6.3781e+06 m           | R_earth, Rearth       ,
-    erg          | 1e-07 kg m2 / s2       |                       ,
+    AU           | 1.49598e+11 m          | au, astronomical_unit            ,
```
?

mhvk replied:
I just copy & pasted this. I think the main change is that this expands it to encompass the long jupter radius entry.

nstarman reviewed (approved):
Overall I think this is a significant improvement!

mhvk commented:
Rebased this PR that removes extra sorting when units are represented as strings with `unit.to_string()`. In general, this seems to improve the readability of the units (especially FITS and VOUNIT), but it implies a lot of small changes (and will probably cause downstream doctests to fail). Two questions:
1. Does this need a what's new entry?
2. I wondered a bit whether one could go for a "standard" representation for anything for which we have a physical type (e.g., "erg/(s cm2 A)" rather than "erg/(A s cm2)" as it is now). Not totally trivial since the physical types do not store how their unit is represented currently. Overall, I felt this might be a good idea but is more work than I have time for now. Of course, it is supplementary to this PR, so can be done after. But ideally we change the format very rarely, so if we merge the present PR, further changes should preferably happen before 5.3 is released.

nstarman commented:
> 1. Does this need a what's new entry?

I think so, esp if point 2 isn't completed by v5.3!

> I wondered a bit whether one could go for a "standard" representation for anything for which we have a physical type (e.g., "erg/(s cm2 A)" rather than "erg/(A s cm2)" as it is now).

I think that would be fantastic!

WilliamJamieson reviewed (approved):
Modeling changes look good to me

mhvk commented:
Thanks! I've added a what's-new entry (and rebased for good measure). It is probably not a bad idea to merge this well before the feature freeze date, so we can get some feedback (e.g., if many doctests break). That also gives some time to experiment with possibly using a default representation based on physical type.

pllim commented:
Need to fix doc warning:

```
docs/whatsnew/5.3.rst:156: WARNING: Duplicate explicit target name: "whatsnew-5.3-unit-formats-fraction".
```

pllim commented on docs/whatsnew/5.3.rst:
```diff
@@ -152,6 +153,21 @@ readable results when printing quantities, table headers and cells, etc.
 For ``'latex'`` the default remains ``fraction='display'``, for an
 unchanged experience with IPython notebook.
 
+.. _whatsnew-5.3-unit-formats-fraction:
```
```suggestion
.. _whatsnew-5.3-unit-formats-order:
```

mhvk replied:
Oops, thanks! Fixed now (I think)

mhvk commented:
OK, tests now pass. @saimn, do you have any worries about the unit changes? @olebole - what do you think?

olebole reviewed (approved):
I am very happy with the changes, as it makes the units more readable by *removing* some code :-) 

What is a bit unfortunate is that it may break unit (doc-) tests of dependent packages, and it would be very hard to formulate the tests in a way that make the tests independent of the Astropy version. This was already true for the other unit format changes; however it reminds that *if* there are other ideas about unit formatting, they should also be implemented before the next Astropy release.

saimn commented:
Indeed the main worry is the impact on unit tests and doctests, but I think it's worth doing.
Maybe add some explanation in https://docs.astropy.org/en/latest/units/format.html as well ?

mhvk commented:
> Maybe add some explanation in https://docs.astropy.org/en/latest/units/format.html as well ?

Hmm, not sure what to do; who would this target?

pllim commented:
How about a new config to retain the old ordering, so Ole's tests won't break? Then eventually, we deprecate that config and then remove it. Or is that just kicking the can down the road with little gain?

nstarman commented:
What about adding a test directive to pytest-astropy, like `FLOAT_CMP` in pytest-doctestplus, that fixes unit comparisons? we could offer that directive for a while and then make it a no-op (equiv to deprecating the config) and then remove it.

pllim commented:
That would make `pytest-doctestplus` be dependent on astropy, introducing circular dependency.

mhvk commented:
I think we are all worrying about something that may be a problem, but perhaps isn't...  So, perhaps merge and see if we get a lot of downstream reports (for those who test against -dev, of course). (I do like the `UNIT_CMP` one, but it is non-trivial even apart from circular dependencies - how does one even recognize a piece of text is a unit?)

pllim commented:
Let's merge then. Can't be worse than that time when we deprecated `clobber` in `astropy.io.fits`. 😉 Thanks, all!

mhvk commented:
Ah, I see. Yes, makes some sense, though maybe even more if the user could control it!

mhvk commented:
But I'll leave this for another time -- nearing end of term here and lots of faculty job candidates coming by...

pllim commented:
So, here are some example failures caused by this PR downstream:

* https://github.com/spacetelescope/jdaviz/issues/2076

What would be the best way to fix these?

mhvk commented:
Sadly, recopy the test string... or write a regex that doesn't check the unit explicitly, or check explicitly where units are the same rather than their representation.

The doctests are the ones that bug me most, as there it is much harder to write it in such a way that the documentation remains legible (so no `...` and `+ELLIPSIS`).

pllim commented:
Here are doctest failures from `specutils` (a Coordinated package). Saw this while working on https://github.com/astropy/specutils/pull/1032

```
____________________________ [doctest] analysis.rst ____________________________
086     known level (defaulting to 1, meaning continuum-normalized).
087     :ref:`specutils-continuum-fitting` describes how continuua can be generated
088     to prepare a spectrum for use with these functions.
089 
090 .. code-block:: python
091 
092     >>> from specutils.analysis import line_flux
093     >>> line_flux(noisy_gaussian, SpectralRegion(7*u.GHz, 3*u.GHz))  # doctest:+FLOAT_CMP
094     <Quantity 4.93784874 GHz Jy>
095     >>> line_flux(noisy_gaussian).to(u.erg * u.cm**-2 * u.s**-1)  # doctest:+FLOAT_CMP
Expected:
    <Quantity 4.97826284e-14 erg / (cm2 s)>
Got:
    <Quantity 4.97951087e-14 erg / (s cm2)>

specutils/docs/analysis.rst:95: DocTestFailure
_________________________ [doctest] spectral_cube.rst __________________________
042 
043     >>> sc.shape # doctest: +REMOTE_DATA
044     (74, 74, 4563)
045 
046 
047 Print the contents of 3 spectral axis points in a 3x3 spaxel array:
048 
049 .. code-block:: python
050 
051     >>> sc[30:33,30:33,2000:2003] # doctest: +REMOTE_DATA
Differences (unified diff with -expected +actual):
    @@ -9,5 +9,5 @@
                [[0.5599331 , 0.554316  , 0.5618426 ],
                 [0.5763055 , 0.5668046 , 0.5774939 ],
    -            [0.59571505, 0.60118765, 0.59942234]]] 1e-17 erg / (Angstrom cm2 s spaxel)>, spectral_axis=<SpectralAxis
    +            [0.59571505, 0.60118765, 0.59942234]]] 1e-17 erg / (Angstrom s spaxel cm2)>, spectral_axis=<SpectralAxis 
        (observer to target:
           radial_velocity=0.0 km / s

specutils/docs/spectral_cube.rst:51: DocTestFailure
```

pllim commented:
Options:

* `# doctest: +SKIP` would allow doc to render without change nor failure, but also disables the doctest.
* `# doctest: +ELLIPSIS` would allow partial doctest that would look a little ugly when rendered and not able to fully test the unit output.
* Using if-else to run two versions of the doc based on astropy version would allow full doctest but might confuse user.

mhvk commented:
Yes, it really is a pain when the format changes. I hadn't quite thought through the problem of different astropy versions that one may be testing against; when numpy changed their format, it was less of a problem since it mostly involved spaces and numbers of digits, which `FLOAT_CMP` could take take of.

Perhaps for `format='general'` we need to revert this until we have a doctest plugin that can deal with this? I'd still like to keep this for `fraction=False`, which affects FITS in particular.

pllim commented:
I don't know. I too agree that this PR is good improvement. I need to sleep on this over the weekend.

nstarman commented:
What about the `UNIT_CMP` flag that can be turned on globally. It wouldn't require Astropy to be installed since it would be a regex to find `<Quantity` and then parse to find the unit bit and then rearrange based on the new rules. Ugh sounds like a fair bit of work, but I think it would solve all pain points except writing it.

pllim commented:
> .. doctest-requires : astropy>=5.3

Unfortunately, no... https://github.com/astropy/pytest-doctestplus/issues/193

But somehow you can do it via `doctest_subpackage_requires` which isn't as flexible but might be better than nothing.

pllim commented:
Yes, but alas you cannot just ask doctestplus to only require certain astropy version for that one line, it is whole module. Not ideal but might be the best we can do for now.

mhvk commented:
I don't think that is that bad -- it shows that examples work with the most recent version of astropy; presumably, the tests cover the rests.

mhvk commented:
Hey, so one can do `__doctest_requires__ = {"BlackBody": ["astropy<5.3"]}`? That's pretty nice at least for the python files! And I think the `.rst` files being covered in `setup.cfg` is not that bad either.

So, I guess the to-do item may be to have this as part of either the what's-new or the API change fragment...

pllim commented:
Might be easier to explain in What's New. Do you want to do a follow-up PR? If not, I can take a stab when I have time.

pllim commented:
Please see https://github.com/astropy/astropy/pull/14526

bsipocz commented:
This bit astroquery as well, and the suggested workaround in setup.cfg is not exactly nice (we do want to test both old and new astropy versions).

mhvk commented:
@bsipocz - sorry to hear that this bit you... I love the principle of doctests, of ensuring that what the user will see when they run some example is actually what is in the documentation, but it does seem they raise the barrier rather a lot for making changes to how output looks - one shouldn't have to think of formatting as part of the API!

bsipocz commented:
OTOH, if formatting is not important enough to consider backward compatibility, it may not be worth changing it and causing such widespread issues downstream, at least not without also adding tools that provide reasonable workarounds (globally skipping doctesting in rst files is not really a nice workaround).

mhvk commented:
It's a hard one - the first unit of `beam^-1 Jy` really had to go; I guess we could have left it at that and minimized the pain by not changing `generic` as well.

On balance, we felt it was worth it. It is a bit like numpy in 1.13 getting around to finally clean-up the number of significant digits that get included in output (following python doing the same). Output looks much better now but it was a gigantic pain (even with `FLOAT_CMP`; we thought about an equivalent for units, but it is just impossible).

WilliamJamieson commented:
Unit string changes will break ASDF unit validation as jsonschema only exactly matches the listed unit string. Resulting in the need for spacetelescope/roman_datamodels#258. I think this maybe the only project currently using ASDF's unit support, so this should not be a big issue right now. In the future, please tag me on any unit string changes.

pllim commented:
@WilliamJamieson , you can always add yourself to the "units" section in codeowners, then you will get pinged automatically. FYI.

mhvk commented:
@WilliamJamieson - I think this is going to be the last string change in a *very long time*.... Hopefully mostly because it is now as "right" as it can be, but certainly also to avoid again causing breakage and thus loss of time that could be better spent elsewhere...
